pho.1
.TH PHO l "March 5 2007" "PHO"
.SH NAME
\fBpho\fP \- View and annotate multiple images
.SH SYNTAX
.B pho
.RI [ options ]
.RI [ filename [ filename... ]]
.SH DESCRIPTION
.I pho
displays images, in the formats handled by the
.IR gdk-pixbuf
package, to an X display.
It is intended as a lightweight and fast viewer,
optimized for rapidly going through large numbers of uploaded images.
.PP
.I pho
is entirely keyboard driven,
and allows for interactive rotation and taking of notes on each image.
It resizes after rotations and will always attempt to show
the image as large as possible.  The final rotation reached for
each image will be remembered and printed when the program exits
(for use with a batch image rotation script).
.PP
.I pho
can also remember up to ten lists of images (numbered 0-9) which can
correspond to anything the user wishes. The image lists will be printed
to standard output when pho exits. Use this to keep notes on which
images you want to save to the web, which images contain images
of your dog, etc.
.SH COMMAND-LINE OPTIONS
.TP
\fB\-p\fR
"Presentation mode": display in full screen mode with a black background
(if the window manager allows it).
.TP
\fB\-P\fR
Force non-presentation mode (e.g. if you set PHO_ARGS=-p).
.TP
\fB\-n\fR
Replace each image with a new window
(slower, but helpful for some broken window managers).
.TP
\fB\-k\fR
Keywords mode: display images scaled down, and show a Keywords dialog
for classifying images. (Keywords mode will cancel Presentation mode.)
.TP
\fB\-sN\fR
Automatic Slideshow mode, where N is the delay in seconds.
For example, -s5 will show pause 5 seconds between images.
-s0 means no delay.
.TP
\fB\-d\fR
Debug mode: may print a few debugging messages to standard output.
.TP
\fB\-h\fR
Help: print a usage statement.
\fB\-v\fR
Verbose help: print a summary of key bindings.
.SH ENVIRONMENT VARIABLES
.TP
PHO_ARGS: default flag settings
(e.g. set it to -p to use presentation mode by default).
Any flags given on the command line will override these settings.
.TP
PHO_REMOTE: the command to call when you press the 'g' key.
Include a %s to represent the filename of the current image.
(Defaults to gimp %s).
.SH KEY BINDINGS
When pho is running, it obeys the following keys:
.TP
\fB[space]\fR
Go to next image. Or cancel slideshow mode, if active.
.TP
\fB[backspace]\fR
Go to previous image
.TP
\fBr\fR, \fBt\fR, \fB[right-arrow]\fR
Rotate right (clockwise)
.TP
\fBR\fR, \fBT\fR, \fBl\fR, \fBL\fR, \fB[left-arrow]\fR
Rotate left (counter-clockwise)
.TP
\fB[up-arrow]\fR, \fB[down-arrow]\fR
Rotate 180 degrees.
.TP
\fBHome\fR
Go back to the first image.
.TP
\fBd\fR
Delete (will bring up a confirmation dialog; clicking OK or
typing another d deletes the file).
.TP
\fBi\fR
Show a dialog with information about the image, including EXIF tags.
.TP
\fB0\fR through \fB9\fR
Add the image to the appropriate notes list
.TP
\fBf\fR
Toggle in/out of "full size mode".  Images will be shown at their
native size, even if it's bigger than the screen size.
(Hint: Many window managers let you move oversized windows with alt-drag.)
.TP
\fBF\fR
Toggle in/out of "full screen mode" (or "fit to window").
Images will be scaled up or down to fill the screen in at least one dimension.
.TP
\fBp\fR
Toggle in/out of "presentation mode".
If the window manager permits, pho will take up the full screen
with the image (if smaller) centered.
.TP
\fB+\fR, \fB=\fR Magnify: show the image at twice the current size.
.TP
\fB/\fR, \fB-\fR Unmagnify: show the image at half the current size.
.TP
\fBq\fR
Quit
.SH ENVIRONMENT
Pho checks for the environment variable PHO_ARGS, which can contain flags
to use as default arguments.
For example, PHO_ARGS=p will always run pho in presentation mode (unless
you pass -P to turn off that mode).
.SH AUTHOR
Akkana Peck
.SH COPYRIGHT
Copyright \(co 2002 - 2007 Akkana Peck
.br
Pho is free software, licensed under the GNU Public License.
.SH SEE ALSO
The most recent version of \fIpho\fR, and more information about it, is at:
http://shallowsky.com/software/pho/


ChangeLog
/2009, Pho 0.9.7-xxx:
    Add ToggleBetween to better manage transitions between various modes
    and make sure proper scaling and redraw happens.

12/15/2009, Pho 0.9.7-pre3:
    More cleanup of rotation/scaling, and transitions from one mode
    to the next.

9/26/2009, Pho 0.9.7-pre2:
    Add ability to add new image files or change the list without restarting.
    Total cleanup/rewrite of rotate/scale logic.
    Remove old gtk1 code.

N/N/2009, Pho 0.9.7-pre1:
    List images with incorrect exif rotation;
    smarter PHO_REMOTE (can do other things, not just GIMP);
    don't go fullscreen from Ctrl-F in the keywords dialog.
    fix a minor Mac build issue.

11/9/2008, Pho 0.9.6:
    RELEASE! The only difference from 0.9.6-pre3 is to change
    the default PHO_REMOTE command from gimp-remote to gimp,
    since gimp-remote is obsolete and no longer provided in Linux
    distros (and never was, on Mac or Win).

10/2/2008, Pho 0.9.6-pre3:
    Change the list next/prev code to fix (I hope) a bug where pho
    would get confused and cycle from the last image back to the first.

8/13/2008, Pho 0.9.6-pre2:
    Simplify exif/Makefile so it'll work on FreeBSD and other
    systems where make is non-GNU make

N/N/2007, Pho 0.9.6-pre1:
    Add new Keywords mode and keywords dialog.
    Add stay-on-top code for prompt and keywords dialogs.
    Clean up a lot of spurious expose events that gtk was sending,
    in the hopes of curing the double-draws visible on slow machines.
    (There are still some a few left, from resize events.)
    A few minor fixes to work better with xinerama.

8/6/2007, Pho 0.9.5.1:
    Fix a buffer overflow bug on the 'g' (gimp-remote) key.

3/20/2007, Pho 0.9.5:
    No changes from pre5 except some manual and html page updates.

3/5/2007, Pho 0.9.5-pre5:
    Check for a PHO_ARGS environment variable. Add a -P argument to force non-presentation mode. Fix a few bugs with double/half size and with scaling.
    Make note 0 work properly. 

7/24/2005, Pho 0.9.5-pre4:
    Double and half size weren't working properly;
    couldn't use arrow and spacebar to get out of the final Quit confirmation dialog (though ESC always worked). 

10/9/2004, Pho 0.9.5-pre3:
    Fixed a crash bug when revisiting an image which was previously rotated.
    Added a workaround for Metacity's evil window size handling.
    Made the background black when in presentation mode. 

9/30/2004, Pho 0.9.5-pre2:
    Fixed some bugs involving ending up at the wrong image when deleting or pressing HOME. Cleaned up handing of image list.
    Fixed an aspect ratio problem. 

7/13/2004, Pho 0.9.5-pre1:
    Rewrote the image list handling to use a linked list. Lots of associated code cleanup.
    Port to gtk2, and build under either gtk1 or gtk2.
    Add new "fullscreen mode" (e.g. for presentations, or for difficult window managers).
    Various futile attempts to hang on to window focus in various window managers. 

9/30/2003, Pho 0.9.2:
    Don't try to move the window unless the image size has changed (fix the annoying bug where pho wouldn't let you move windows).
    Also fix a crash after dismissing the info dialog by clicking on the windowmanager "X". 

2/26/2003, Pho 0.9.1:
    Rusty Russell: fix an off-by-one error in rotation code, reducing the need for extra memory;
    Brian Langenberger: bracket printed filenames with quotes if they contain spaces or quote characters. 

1/8/2003, Pho 0.9:
    Fix pho.spec version, and a make clean bug. 

12/3/2002, Pho 0.9pre2:
    Fix a crash in the quit dialog. 

11/27/2002, Pho 0.9pre1:
    Add EXIF handling, using code from jhead, by Matthias Wandel. 

10/18/2002, Pho 0.8:
    Build RPM and deb; show image size in titlebar; add home key. 

8/13/2002, Pho 0.7:
    Put image name in titlebar; prompt after last image before quitting; add fullscreen mode. 

7/28/2002, Pho 0.6.1:
    Fixed a bug with printing a note set on the last image. 

7/21/2002, Pho 0.6:
    Code cleanup, change the handling of the notes structures. 

7/21/2002, Pho 0.5.1:
    Fix a bug where image rotations were wrong when going backward. 

7/11/2002, Pho 0.5:
    Add delete dialog, and really delete files on disk. 

7/3/2002, Pho 0.4:
    Fixed a bug with gif rotation. 

7/1/2002, Pho 0.3:
    Rename yass to pho, add info dialog. 

6/28/2002, Yass 0.2

pho.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * pho.c: core routines for pho, an image viewer.
 *
 * Copyright 2004 by Akkana Peck.
 * You are free to use or modify this code under the Gnu Public License.
 */

#include "pho.h"
#include "dialogs.h"
#include "exif/phoexif.h"

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>    /* for unlink() */

/* ************* Definition of globals ************ */
PhoImage* gFirstImage = 0;
PhoImage* gCurImage = 0;

/* Monitor resolution */
int gMonitorWidth = 0;
int gMonitorHeight = 0;

/* We only have one image at a time, so make it global. */
GdkPixbuf* gImage = 0;

int gDebug = 0;    /* debugging messages */

int gScaleMode = PHO_SCALE_NORMAL;
double gScaleRatio = 1.0;

int gDisplayMode = PHO_DISPLAY_NORMAL;

/* Slideshow delay is zero by default -- no slideshow. */
int gDelaySeconds = 0;
int gPendingTimeout = 0;

static int RotateImage(PhoImage* img, int degrees);    /* forward */

static gint DelayTimer(gpointer data)
{
    if (gDelaySeconds == 0)    /* slideshow mode was cancelled */
        return FALSE;

    if (gDebug) printf("-- Timer fired\n");
    gPendingTimeout = 0;

    NextImage();
    return FALSE;       /* cancel the timer */
}

int ShowImage()
{
    ScaleAndRotate(gCurImage, 0);
    /* Keywords dialog will be updated if necessary from DrawImage */

    if (gDelaySeconds > 0 && gPendingTimeout == 0
        && (gCurImage->next != 0 || gCurImage->next != gFirstImage)) {
        if (gDebug) printf("Adding timeout for %d msec\n", gDelaySeconds*1000);
        gPendingTimeout = g_timeout_add (gDelaySeconds * 1000, DelayTimer, 0);
    }

    return 0;
}

static int LoadImageFromFile(PhoImage* img)
{
    GError* err = NULL;
    int rot;

    if (img == 0)
        return -1;

    if (gDebug)
        printf("LoadImageFromFile(%s)\n", img->filename);

    /* Free the current image */
    if (gImage) {
        gdk_pixbuf_unref(gImage);
        gImage = 0;
    }

    gImage = gdk_pixbuf_new_from_file(img->filename, &err);
    if (!gImage)
    {
        gImage = 0;
        fprintf(stderr, "Can't open %s: %s\n", img->filename, err->message);
        return -1;
    }

    img->curWidth = gdk_pixbuf_get_width(gImage);
    img->curHeight = gdk_pixbuf_get_height(gImage);

    /* The first time an image is loaded, it should be rotated
     * to its appropriate EXIF rotation. Subsequently, though,
     * it should be rotated to curRot.
     */
    if (img->trueWidth == 0 || img->trueHeight == 0) {
        /* Read the EXIF rotation if we haven't already rotated this image */
        ExifReadInfo(img->filename);
        if (HasExif() && (rot = ExifGetInt(ExifOrientation)) != 0)
            img->exifRot = rot;
        else
            img->exifRot = 0;
    }

    /* trueWidth and Height used to be set inside EXIF clause,
     * but that doesn't make sense -- we need it not just the first
     * time, but also ever time the image is reloaded.
     */
    img->trueWidth = img->curWidth;
    img->trueHeight = img->curHeight;

    return 0;
}

static int LoadImageAndRotate(PhoImage* img)
{
    int e;
    int rot = img->curRot;
    int firsttime = (img->trueWidth == 0);

    img->trueWidth = img->trueHeight = img->curRot = 0;

    e = LoadImageFromFile(img);
    if (e) return e;

    /* If it's not the first time we've loaded this image,
     * default its rotation to the EXIF rotation if any.
     * Otherwise rotate to the saved img->curRot.
     */
    if (firsttime && img->exifRot != 0)
        ScaleAndRotate(gCurImage, img->exifRot);

    else
        ScaleAndRotate(gCurImage, rot);

    return 0;
}

/* ThisImage() is called when gCurImage has changed and needs to
 * be reloaded.
 */
int ThisImage()
{
    if (LoadImageAndRotate(gCurImage) != 0)
        return NextImage();
    ShowImage();
    return 0;
}

int NextImage()
{
    if (gDebug)
        printf("\n================= NextImage ====================\n");
    do {
        if (gCurImage == 0)   /* no image loaded yet, first call */
            gCurImage = gFirstImage;
        else if ((gCurImage->next == 0) || (gCurImage->next == gFirstImage))
            return -1;  /* end of list, can't go farther */
        else
            gCurImage = gCurImage->next;
    } while (LoadImageAndRotate(gCurImage) != 0);
    ShowImage();
    return 0;
}

int PrevImage()
{
    if (gDebug)
        printf("\n================= PrevImage ====================\n");
    do {
        if (gCurImage == 0) {  /* no image loaded yet, first call */
            gCurImage = gFirstImage;
            if (gCurImage->prev)
                gCurImage = gCurImage->prev;
        }
        else {
            if (gCurImage == gFirstImage)
                return -1;  /* end of list */
            gCurImage = gCurImage->prev;
        }
    } while (LoadImageAndRotate(gCurImage) != 0);
    ShowImage();
    return 0;
}

/* Limit new_width and new_height so that they're no bigger than
 * max_width and max_height.
 */
static void ScaleToFit(int *new_width, int *new_height,
                       int max_width, int max_height)
{
    /* scale so that the biggest ratio just barely fits on screen. */
    double xratio = (double)max_width / *new_width;
    double yratio = (double)max_height / *new_height;
    double ratio = MIN(xratio, yratio);

    *new_width = ratio * *new_width;
    *new_height = ratio * *new_height;
}

#define SWAP(a, b) { int temp = a; a = b; b = temp; }
/*#define SWAP(a, b)  {a ^= b; b ^= a; a ^= b;}*/

/* Rotate the image according to the current scale mode, scaling as needed,
 * then redisplay.
 * 
 * This will read the image from disk if necessary,
 * and it will rotate the image at the appropriate time
 * (when the image is at its smallest).
 *
 * This is the routine that should be called by external callers:
 * callers should never need to call RotateImage.
 *
 * degrees is the increment from the current rotation (curRot).
 */
void ScaleAndRotate(PhoImage* img, int degrees)
{
#define true_width img->trueWidth
#define true_height img->trueHeight
    int new_width;
    int new_height;

    if (gDebug)
        printf("ScaleAndRotate(%d (cur = %d))\n", degrees, img->curRot);

    /* degrees should be between 0 and 360 */
    degrees = (degrees + 360) % 360;

    /* First, load the image if we haven't already, to get true w/h */
    if (true_width == 0 || true_height == 0) {
        if (gDebug) printf("Loading, first time, from ScaleAndRotate!\n");
        LoadImageFromFile(img);
    }

    /*
     * Calculate new_width and new_height, the size to which the image
     * should be scaled before or after rotation,
     * based on the current scale mode.
     * That means that if the aspect ratio is changing,
     * new_width will be the image's height after rotation.
     */

    /* Fullsize: display always at real resolution,
     * even if it's too big to fit on the screen.
     */
    if (gScaleMode == PHO_SCALE_FULLSIZE) {
        new_width = true_width * gScaleRatio;
        new_height = true_height * gScaleRatio;
        if (gDebug) printf("Now fullsize, %dx%d\n", new_width, new_height);
    }

    /* Normal: display at full size unless it won't fit the screen,
     * in which case scale it down.
     */
#define NORMAL_SCALE_SLOP 5
    else if (gScaleMode == PHO_SCALE_NORMAL
             || gScaleMode == PHO_SCALE_SCREEN_RATIO) {
        int max_width, max_height;
        int aspect_changing;    /* Is the aspect ratio changing? */

        new_width = true_width;
        new_height = true_height;

        aspect_changing = ((degrees % 180) != 0);
        if (aspect_changing) {
            max_width = gMonitorHeight;
            max_height = gMonitorWidth;
            if (gDebug)
                printf("Aspect ratio is changing\n");
        } else {
            max_width = gMonitorWidth;
            max_height = gMonitorHeight;
        }
        if (new_width > max_width || new_height > max_height)
            ScaleToFit(&new_width, &new_height, max_width, max_height);

        /* In case we're scaling: */
        new_width *= gScaleRatio;
        new_height *= gScaleRatio;

        /* See if new_width and new_height are close enough already
         * that it might not be worth doing the work of scaling:
         */
        if (abs(img->curWidth - new_width) + abs(img->curHeight - new_height)
            < NORMAL_SCALE_SLOP) {
            new_width = img->curWidth;
            new_height = img->curHeight;
        }
    }

    else if (gScaleMode == PHO_SCALE_IMG_RATIO) {
        new_width = true_width * gScaleRatio;
        new_height = true_height * gScaleRatio;

        /* See if we're close */
        if (abs(img->curWidth - new_width) + abs(img->curHeight - new_height)
            < NORMAL_SCALE_SLOP) {
            new_width = img->curWidth;
            new_height = img->curHeight;
        }
    }
        
    /* Fullscreen: Scale either up or down if necessary to make
     * the largest dimension match the screen size.
     */
#define FULLSCREEN_SCALE_SLOP 20
    else if (gScaleMode == PHO_SCALE_FULLSCREEN) {
        /* If we're in presentation mode, then we need to scale
         * to the current size of the window, not the monitor,
         * because in xinerama gdk_window_fullscreen() will fullscreen
         * onto only one monitor, but gdk_screen_width() gives the
         * width of the full xinerama setup.
         */
        int diffx = abs(img->curWidth - gMonitorWidth);
        int diffy = abs(img->curHeight - gMonitorHeight);
        double xratio, yratio;
        gint screenwidth, screenheight;

        if (gDisplayMode == PHO_DISPLAY_PRESENTATION)
            gtk_window_get_size(GTK_WINDOW(gWin), &screenwidth, &screenheight);
        else {
            screenwidth = gMonitorWidth;
            screenheight = gMonitorHeight;
        }

        if (diffx < FULLSCREEN_SCALE_SLOP || diffy < FULLSCREEN_SCALE_SLOP) {
            new_width = img->curWidth;
            new_height = img->curHeight;
        }
        else {
            xratio = (double)screenwidth / true_width;
            yratio = (double)screenheight / true_height;

            /* Use xratio for the more extreme of the two */
            if (xratio > yratio) xratio = yratio;
            new_width = xratio * true_width;
            new_height = xratio * true_height;
        }
    }
    else {
        /* Shouldn't ever happen, means gScaleMode is bogus */
        printf("Internal error: Unknown scale mode %d\n", gScaleMode);
        new_width = img->curWidth;
        new_height = img->curHeight;
    }
    /*
     * Finally, we're done with the scaling modes.
     * Time to do the scaling and rotation,
     * reloading the image if needed.
     */

    /* First figure out if we're getting bigger and hence need to reload. */
    if ((new_width > img->curWidth || new_height > img->curHeight)
        && (img->curWidth < true_width && img->curHeight < true_height)) {
        if (gDebug)
            printf("Getting bigger, from %dx%d to %dx%d -- need to reload\n",
                   img->curWidth, img->curHeight, new_width, new_height);

        /* Because curRot is going back to zero, that means we
         * might need to swap new_width and new_height, in case
         * the aspect ratio is changing
        //if (img->curRot % 180 != 0) {
         */
        if (degrees % 180 != 0) {
            SWAP(new_width, new_height);
        }

        /* image->curRot is going to be set back to zero when we load
         * from file, so adjust current planned rotation accordingly.
         */
        degrees = (degrees + img->curRot + 360) % 360;
            /* Now it's the absolute end rot desired */

        img->curRot = 0;
        LoadImageFromFile(img);
    }
#if 0
    else if (degrees % 180 != 0) {
        SWAP(new_width, new_height);
        printf("Not reloading, but swapped new width/height, now %dx%d\n",
               new_width, new_height);
    }

    /* new_* are the sizes we want after rotation. But we need
     * to compare them now to the sizes before rotation. So we
     * need to compensate for rotation.
     */

    /* If we're going to scale up, then do the rotation first,
     * before scaling. Otherwise, scale down first then rotate.
     */
    if (degrees != 0 &&
        (new_width > img->curWidth || new_height > img->curHeight)) {
        if (gDebug) printf("Rotating before scaling up\n");
        printf("Will be scaling from %dx%d to %dx%d\n",
               img->curWidth, img->curHeight, new_width, new_height);
        RotateImage(img, degrees);
        degrees = 0;    /* finished with rotation */
    }
#endif

    /* Do the scaling (thought we'd never get there!) */
    if (new_width != img->curWidth || new_height != img->curHeight)
    {
        GdkPixbuf* newimage = gdk_pixbuf_scale_simple(gImage,
                                                      new_width,
                                                      new_height,
                                                      GDK_INTERP_BILINEAR);
        /* If that's too slow use GDK_INTERP_NEAREST */

        /* scale_simple apparently has no error return; if it fails,
         * it still returns a pixbuf but width and height are -1.
         * Hope this undocumented behavior doesn't change!
         * Later: now it's documented. Whew.
         */
        if (!newimage || gdk_pixbuf_get_width(newimage) < 1) {
            printf("\007Error scaling up to %d x %d: probably out of memory\n",
                   new_width, new_height);
            if (newimage)
                gdk_pixbuf_unref(newimage);
            Prompt("Couldn't scale up: probably out of memory", "Bummer", 0,
                   "\n ", "");
            return;
        }
        if (gDebug)
            printf("Scale from %dx%d = %dx%d to %dx%d = %dx%d\n",
                   img->curWidth, img->curHeight,
                   gdk_pixbuf_get_width(gImage), gdk_pixbuf_get_height(gImage),
                   new_width, new_height,
                   gdk_pixbuf_get_width(newimage),
                   gdk_pixbuf_get_height(newimage));
        if (gImage)
            gdk_pixbuf_unref(gImage);
        gImage = newimage;

        img->curWidth = gdk_pixbuf_get_width(gImage);
        img->curHeight = gdk_pixbuf_get_height(gImage);
    }

    /* If we didn't rotate before, do it now. */
    if (degrees != 0)
        RotateImage(img, degrees);

    /* We've finished making our changes. Now we may need to make
     * changes in the window size or position.
     */
    PrepareWindow();
}

PhoImage* NewPhoImage(char* fnam)
{
    PhoImage* newimg = calloc(1, sizeof (PhoImage));
    if (newimg == 0) return 0;
    newimg->filename = fnam;  /* no copy, we don't own the memory */

    return newimg;
}

/* This routine exists to keep track of any allocated memory
 * existing in the PhoImage structure.
 */
static void FreePhoImage(PhoImage* img)
{
    if (img->comment) free(img->comment);
    free(img);
}

/* Remove all images from the image list, to start fresh. */
void ClearImageList()
{
    PhoImage* img = gFirstImage;
    do {
        PhoImage* next = img->next;
        FreePhoImage(img);
        img = next;
    } while (img && img != gFirstImage);

    gCurImage = gFirstImage = 0;
}

void ReallyDelete(PhoImage* delImg)
{
    if (unlink(delImg->filename) < 0)
    {
        printf("OOPS!  Can't delete %s\n", delImg->filename);
        return;
    }

    /* next and prev should never be 0, but check anyway */
    if (delImg->next == 0)
        printf("BUG: delImg->next is zero!\n");
    if (delImg->prev == 0)
        printf("BUG: delImg->prev is zero!\n");

    /* Disconnect it from the linked list, and reset gCurImage. */
    if (delImg == gFirstImage                /* this is the only image! */
        && (delImg->next == gFirstImage || delImg->next == 0)) {
        EndSession();
    }
    else if (delImg->prev == delImg->next) { /* One image left after this */
        gFirstImage = gCurImage = delImg->prev;
        gCurImage->prev = gCurImage->next = gCurImage;
    }
    else if (delImg->next == gFirstImage) {  /* Last image, so go back */
        gCurImage = delImg->prev;
        gCurImage->next = gFirstImage;
        gFirstImage->prev = gCurImage;
    }
    else {
        gCurImage = delImg->next;
        gCurImage->prev = delImg->prev;
        if (delImg->prev == 0) {
            printf("BUG: delImg->prev == 0!\n");
            delImg->prev = gFirstImage;
        }
        delImg->prev->next = gCurImage;
        if (gCurImage->next == 0) {
            printf("BUG: delImg->prev == 0!\n");
            gCurImage->next = gFirstImage;
        }
        gCurImage->next->prev = gCurImage;
    }
    /* If we deleted the first image, make sure we reset the list */
    if (delImg == gFirstImage) {
        gFirstImage = delImg->next;
    }

    /* It's disconnected.  Free all the memory */
    FreePhoImage(delImg);

    ThisImage();
}

void DeleteImage(PhoImage* delImg)
{
    char buf[512];
    if (delImg->filename == 0)
        return;
    sprintf(buf, "Delete file %s?", delImg->filename);
    if (Prompt(buf, "Delete", 0, "dD\n", "nN") > 0)
        ReallyDelete(delImg);
}

/* RotateImage just rotates an existing image, no scaling or reloading.
 * It's typically called from ScaleAndRotate either just
 * before or just after scaling.
 * No one except ScaleAndRotate should call it.
 * Degrees is the amount of rotation relative to current.
 */
static int RotateImage(PhoImage* img, int degrees)
{
    guchar *oldpixels, *newpixels;
    int x, y;
    int oldrowstride, newrowstride, nchannels, bitsper, alpha;
    int newWidth, newHeight, newTrueWidth, newTrueHeight;
    GdkPixbuf* newImage;

    if (!gImage) return 1;     /* sanity check */

    if (gDebug)
        printf("RotateImage(%d), initially %d x %d, true %dx%d\n",
               degrees, img->curWidth, img->curHeight,
               img->trueWidth, img->trueHeight);

    /* Make sure degrees is between 0 and 360 even if it's -90 */
    degrees = (degrees + 360) % 360;

    /* degrees might be zero now, since we might be rotating back to zero. */
    if (degrees == 0) {
        return 0;
    }

    /* Swap X and Y if appropriate */
    if (degrees == 90 || degrees == 270)
    {
        newWidth = img->curHeight;
        newHeight = img->curWidth;
        newTrueWidth = img->trueHeight;
        newTrueHeight = img->trueWidth;
    }
    else
    {
        newWidth = img->curWidth;
        newHeight = img->curHeight;
        newTrueWidth = img->trueWidth;
        newTrueHeight = img->trueHeight;
    }

    oldrowstride = gdk_pixbuf_get_rowstride(gImage);
    /* Sometimes rowstride is slightly different from width*nchannels:
     * gdk_pixbuf optimizes by aligning to 32-bit boundaries.
     * But apparently it works even if rowstride is not aligned,
     * just might not be as fast.
     * XXX check newrowstride alignment
     */
    bitsper = gdk_pixbuf_get_bits_per_sample(gImage);
    nchannels = gdk_pixbuf_get_n_channels(gImage);
    alpha = gdk_pixbuf_get_has_alpha(gImage);

    oldpixels = gdk_pixbuf_get_pixels(gImage);

    newImage = gdk_pixbuf_new(GDK_COLORSPACE_RGB, alpha, bitsper,
                              newWidth, newHeight);
    if (!newImage) return 1;
    newpixels = gdk_pixbuf_get_pixels(newImage);
    newrowstride = gdk_pixbuf_get_rowstride(newImage);

    for (x = 0; x < img->curWidth; ++x)
    {
        for (y = 0; y < img->curHeight; ++y)
        {
            int newx, newy;
            int i;
            switch (degrees)
            {
              case 90:
                newx = img->curHeight - y - 1;
                newy = x;
                break;
              case 270:
                newx = y;
                newy = img->curWidth - x - 1;
                break;
              case 180:
                newx = img->curWidth - x - 1;
                newy = img->curHeight - y - 1;
                break;
              default:
                printf("Illegal rotation value!\n");
                return 1;
            }
            for (i=0; i<nchannels; ++i)
                newpixels[newy*newrowstride + newx*nchannels + i]
                    = oldpixels[y*oldrowstride + x*nchannels + i];
        }
    }

    img->curWidth = newWidth;
    img->curHeight = newHeight;
    img->trueWidth = newTrueWidth;
    img->trueHeight = newTrueHeight;

    img->curRot = (img->curRot + degrees + 360) % 360;

    gdk_pixbuf_unref(gImage);
    gImage = newImage;

    return 0;
}

void Usage()
{
    printf("pho version %s.  Copyright 2002-2009 Akkana Peck akkana@shallowsky.com.\n", VERSION);
    printf("Usage: pho [-dhnp] image [image ...]\n");
    printf("\t-p: Presentation mode (full screen)\n");
    printf("\t-k: Keywords mode (show a Keywords dialog for each image)\n");
    printf("\t-n: Replace each image window with a new window (helpful for some window managers)\n");
    printf("\t-sN: Slideshow mode, where N is the timeout in seconds\n");
    printf("\t-d: Debug messages\n");
    printf("\t-h: Help: Print this summary\n");
    printf("\t-v: Verbose help: Print a summary of key bindings\n");
    exit(1);
}

void VerboseHelp()
{
    printf("pho version %s.  Copyright 2002,2003,2004,2007 Akkana Peck akkana@shallowsky.com.\n", VERSION);
    printf("Type pho -h for commandline arguments.\n");
    printf("\npho Key Bindings:\n\n");
    printf("<space>\tNext image (or cancel slideshow mode)\n");
    printf("-\tPrevious image\n");
    printf("<backspace>\tPrevious image\n");
    printf("<home>\tFirst image\n");
    printf("f\tToggle fullscreen mode (scale even small images up to fullscreen)\n");
    printf("F\tToggle full-size mode (even if bigger than screen)\n");
    printf("k\tTurn on keywords mode: show the keywords dialog\n");
    printf("p\tToggle presentation mode (take up the whole screen, centering the image)\n");
    printf("d\tDelete current image (from disk, after confirming with another d)\n");
    printf("0-9\tRemember image in note list 0 through 9 (to be printed at exit)\n");
    printf("t\tRotate right 90 degrees\n");
    printf("r\tRotate right 90 degrees\n");
    printf("<Right>\tRotate right 90 degrees\n");
    printf("T\tRotate left 90 degrees\n");
    printf("R\tRotate left 90 degrees\n");
    printf("l\tRotate left 90 degrees\n");
    printf("L\tRotate left 90 degrees\n");
    printf("<left>\tRotate left 90 degrees\n");
    printf("<up>\tRotate 180 degrees\n");
    printf("+\tDouble size\n");
    printf("=\tDouble size\n");
    printf("/\tHalf size\n");
    printf("<kp>-\tHalf size\n");
    printf("i\tShow/hide info dialog\n");
    printf("o\nChange the working file set (add files or make a new list)\n");
    printf("g\tRun gimp on the current image\n");
    printf("\t(or set PHO_REMOTE to an alternate command)\n");
    printf("q\tQuit\n");
    printf("<esc>\tQuit (or hide a dialog, if one is showing)\n");
    exit(1);
}







